### DISCUSSION TITLE
BON8 support

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#2980] [Ideas]
In a recent [Reddit discussion](https://www.reddit.com/r/cpp/comments/p9ftmk/easy_serialization_library/ha0minz/?context=8&depth=9) I learned about the BON8 format which is another binary format for JSON, but with a unique idea to exploit some characteristics of UTF-8 strings.

Maybe this format could be added as prototype to the library.

- [specification](https://github.com/ttauri-project/ttauri/blob/main/docs/BON8.md)
- [optimized specification](https://github.com/ttauri-project/ttauri/blob/audio-enumerate-modes/docs/BON8.md)
- [implementation](https://github.com/ttauri-project/ttauri/blob/main/src/ttauri/codec/BON8.hpp)

/cc @takev

**Comment by takev** [comment]
As discussed in the reddit post, I had updated the BON8 specification to handle small arrays and objects better.

I have now merged those changes to the main branch, and the "specification" link above post to the correct version.
The "optimized specification" will disappear when I am done working on that branch. The "implementation" is still based on the old specification and will be replaced when I merge the branch I am working on.

No more changes will be done to the codes and their meaning.

If there are other bugs in the specification, or things that are unclear, please don't forget to contact me and attempt to fix those issues.

**Reply by nlohmann** [reply]
I have some questions:

1. `0.0` has a compact encoding of `0xFC` - but how is `-0.0` encoded? As `0x8E 80 00 00 00`?
2. How is `{"a": "", "c": "d"}` encoded:  As `0x88 0x61 0xFF 0xFF 0x63 0xFF 0x64`? (I am confused about the rules when to insert a `0xFF`.

**Comment by takev** [comment]
-0.0 should be encoded as 0xfc. Mostly because 0.0 compares equal to -0.0, and has pretty limited impact.

The last "d" needs to be terminated with 0xff as well, due to it being at the end of the message.

Append 0xff when:
 - The string is empty
 - The string is the last thing in the message (also if the string is at the end of the last object or array).
 - The string is followed by another string (also if the string is at the end of an object or array, which is followed by a string).

I will update the documentation to be more clear, maybe add a few examples.

**Reply by nlohmann** [reply]
Thanks for the response!

- I think that dropping the sign of `-0.0` is not a good idea. Both CBOR and MessagePack can round-trip the sign of `-0.0` and BON8 would lose precision here.
- Thanks for clarifying. I indeed forgot the trailing `0xFF` as I was focussing on objects at that point. I think having some examples would be really handy.

**Reply by gregmarr** [reply]
@takev Just so that you don't think that the `-0.0` is a purely theoretical thing, we've had user requests around preserving `-0.0` within the last couple months.  (Edit: a little longer ago than I thought, but not that long.)

**Reply by nlohmann** [reply]
Right, see #2854.

**Reply by takev** [reply]
Oh very interesting. Okay, lets encode -0.0 as 0x8E 80 00 00 00.
It should be rare, so using 5 bytes for that is fine.

I will change the implementation to check for the sign of zero.

**Reply by takev** [reply]
I think I can improve on the description of how string termination works, with a single rule.

1. All strings are terminated using eot (0xff), except when a non-empty string is followed by a byte that is not part of a valid UTF-8 string.

**Reply by takev** [reply]
actually that is still three rules, but now written as a single sentence. hmmm

**Reply by gregmarr** [reply]
Since I had to go look up why it would be encoded as that particular 5 byte sequence, here is that info for anyone else reading this.  `8e` indicates that the next 4 bytes are a `Floating point binary32`, which in this case is `80 00 00 00` or zero with the high bit set.

**Reply by takev** [reply]
I thought I would be a bit more clear on floating point binary32 or binary64 selection. I needed to allow for different rounding modes, which is why the check is within 1ULP, instead of 0ULP.

```cpp
bool binary64_can_be_represented_as_binary32(double x)
{
    if (std::isnan(x) or std::isinf(x)) {
        return true;
    } else if (x < -std::numeric_limits<float>::max() or x > std::numeric_limits<float>::max) {
        return false;
    } else {
        auto y = static_cast<float>(x);
        auto a = std::nextafter(x, -std::numeric_limits<double>::infinity());
        auto b = std::nextafter(x, std::numeric_limits<double>::infinity());
        return a <= y and y <= b;
    }
}
```

**Reply by nlohmann** [reply]
Just out of curiosity: we are using this code to decide whether the number `n` can be represented by a `float`.

```cpp
static_cast<double>(n) >= static_cast<double>(std::numeric_limits<float>::lowest()) &&
    static_cast<double>(n) <= static_cast<double>((std::numeric_limits<float>::max)()) &&
    static_cast<double>(static_cast<float>(n)) == static_cast<double>(n))
```

it's been a while since it was written - are we missing anything here?

**Reply by takev** [reply]
I did some research, it looks like binary32 to binary64 conversion is perfect. Therefor yes we can do an exact comparison.

I also check for binary64 infinite and nan and convert those to binary32 as well.

**Reply by nlohmann** [reply]
I have some more questions/comments about the specification:

- You have a compact 1-byte representation for positive integers < 40. And then there is a 2-byte representation for positive integers < 3839. It is possible to represent the numbers 0..39 in both ways. Of course, the specification demands that numbers should be represented as small as possible - this means that you "waste" 39 2-byte numbers which will never be used. What I mean: instead of the range 0.. 3839 rather have a range 40..3879, meaning `c2 00` to represent 40 rather than 0. The same effect holds for other ranges.
- Why do you demand object keys to be sorted?

**Reply by takev** [reply]
About the integers wasting ranges, this is true. I thought it wouldn't make too much of a difference, and encoding would be simpler as no offset needs to be added or removed.

I based it on how UTF-8 also encodes the code-points starting from zero for each of the four different UTF-8 sequences. And since I mirrored UTF-8 I thought doing the same with integers would reduce confusion.

The keys need to be sorted so that BON8 only has one canonical representation. Meaning all encoders will generate the exact same byte sequence. This would allow signatures to be compared from the original data without having to transfer the BON8 encoded version of the data, maybe that is taking it too far?

**Reply by nlohmann** [reply]
- As for the integers, I think it is rather confusing that you allow for **5** different representations of the numbers `0` to `39`, and then add a rule "Integers MUST be encoded in the least amount of bytes.". I don't think it has a huge impact size-wise, but for my taste, it would feel more elegant to allow for more numbers to be encoded with fewer bytes. And since there's already a lot of shifting, masking, and subtraction going on, I don't think that adding another step to save some bytes would make it too complex. That said, I think you leave room for optimizations unused which would be a shame given the elegance of your approach.
- Thanks for clarifying your thoughts on object key orders. I personally think that demanding a fixed order feels like forcing an implementation detail to encoders just to have some canonic result. I have not seen such a requirement in other binary formats, and I would rather add a comment like "If object keys are ordered lexicographically, the BON8 encoding of a JSON value is unique".

**Reply by takev** [reply]
Okay, I relent.

Integers now have fitting ranges, and canonical representation is now optional.

  Code Sequence           | # | Type                        | Bits |      Min |       Max
 :----------------------- | -:|:-------------------------   | ----:| --------:| ----------:
  90-b7                   | 1 | Positive integer            |      |        0 |        39
  b8-c1                   | 1 | Negative Integer            |      |       -1 |       -10
  c2-df 00-7f             | 2 | Positive Integer (30 * 128) |      |       40 |      3879
  e0-ef 00-7f byte\*1     | 3 | Positive Integer            |   19 |     3880 |    528167
  f0-f7 00-7f byte\*2     | 4 | Positive Integer            |   26 |   528168 |  67637031
  c2-df c0-ff             | 2 | Negative Integer (30 * 64)  |      |      -11 |     -1930
  e0-ef c0-ff byte\*1     | 3 | Negative Integer            |   18 |    -1931 |   -264074
  f0-f7 c0-ff byte\*2     | 4 | Negative Integer            |   25 |  -264075 | -33818506

Extra rules
-----------

The rules below ensures minimum message size and consistency, which allows for cryptographically
signing of messages. All encoders MUST follow these rules.

 - A message is a single value. Most often this value is of type Object or type Array.
 - A string does not need to be terminated with an eos (0xff), if the ending of a string
   can be determined in other ways.
 - String MUST end with eos (0xff) when:
   - The string is empty,
   - When the next byte in the message starts another string,
   - If there are no more bytes left in the message.
 - Strings MUST be a valid UTF-8 encoded string.
 - Strings MAY contain any Unicode code-point between U+0000 and U+10ffff.
 - Integers and floating point numbers are stored most significant bits first (big endian).

Canonical Representation
------------------------
The rules below ensure a minimum message and a consistent encoding between different
implementation. The canonical representation is consistent enough to be used for cryptographically
signing of data.

 - The unicode string MUST be in Normalization Form C (NFC).
 - Floating point negative zero, infinite and NaN MUST be encoded as binary32.
 - NaN should be encoded as a 0x7f800001. 
 - Floating point numbers that can be converted to binary32 without loss
   of precision or range MUST be encoded as binary32.
 - Strings, Integers, Arrays and Objects MUST be encoded with the least amount of bytes.
 - The keys of an Object MUST be lexically ordered based on UTF-8 code-units.
 -

**Reply by nlohmann** [reply]
> Floating point numbers that can be converted to binary32 without loss of precision or range MUST be encoded as binary32.

Do you mean "float32" rather than "binary32"?

**Reply by takev** [reply]
In the IEEE 754-2008 standard, the 32-bit base-2 format is officially referred to as binary32.

**Reply by nlohmann** [reply]
Thanks for adjusting the integer ranges - I just committed the changes to the `bon8` branch. The BON8 writer is nearly finished - there are likely bugs in it, but I would need more examples to find them. Do you have some test suite or list of examples for the corner cases?

And, just to clarify: You write "NaN should be encoded as a 0x7f800001." - I assume `0x7f800001` is a float, so the BON8 representation of NaN is `0x8e7f800001`, right?

**Reply by takev** [reply]
Sorry, a bit late.

I do not have many tests yet. Sadly I have a lot of code to maintain the BON8 specification has a few examples in regard to strings.

The NaN should indeed be encoded as a 0x8e 0x7f 0x80 0x00 0x01. NaNs hopefully are not normal for data, so storing them as 5 bytes is hopefully fine.

**Comment by nlohmann** [comment]
FYI, I started a PR for BON8 support: #2998.
